File: packages/desktop/src/browser/platform/macos_wkwebview.m
#import <AppKit/AppKit.h>
#import <Foundation/Foundation.h>
#import <WebKit/WebKit.h>
#include <stdlib.h>
#include <string.h>

enum VerdeMacBrowserEventKind {
    VERDE_MAC_BROWSER_EVENT_OPENED = 1,
    VERDE_MAC_BROWSER_EVENT_CLOSED = 2,
    VERDE_MAC_BROWSER_EVENT_NAVIGATED = 3,
    VERDE_MAC_BROWSER_EVENT_TITLE_CHANGED = 4,
    VERDE_MAC_BROWSER_EVENT_DOCUMENT_LOADED = 5,
    VERDE_MAC_BROWSER_EVENT_JS_MESSAGE = 6,
    VERDE_MAC_BROWSER_EVENT_EVAL_RESULT = 7,
    VERDE_MAC_BROWSER_EVENT_FAILED = 8,
};

@interface VerdeMacBrowserEvent : NSObject
@property(nonatomic) int kind;
@property(nonatomic, copy, nullable) NSString *payload;
+ (instancetype)eventWithKind:(int)kind payload:(nullable NSString *)payload;
@end

@implementation VerdeMacBrowserEvent
+ (instancetype)eventWithKind:(int)kind payload:(nullable NSString *)payload {
    VerdeMacBrowserEvent *event = [[VerdeMacBrowserEvent alloc] init];
    event.kind = kind;
    event.payload = payload;
    return event;
}
@end

@interface VerdeMacBrowser : NSObject <WKScriptMessageHandler, WKNavigationDelegate>
@property(nonatomic, weak) NSWindow *window;
@property(nonatomic, strong) NSView *container;
@property(nonatomic, strong) WKWebView *webView;
@property(nonatomic, strong) NSMutableArray<VerdeMacBrowserEvent *> *events;
@property(nonatomic) BOOL visible;
@property(nonatomic) BOOL invalidated;
- (instancetype)initWithWindow:(NSWindow *)window;
- (void)invalidate;
- (void)queueEvent:(int)kind payload:(nullable NSString *)payload;
- (void)setScreenX:(int)x y:(int)y width:(int)width height:(int)height;
@end

static NSString *VerdeMacStringFromCString(const char *value) {
    if (value == NULL) return @"";
    return [NSString stringWithUTF8String:value] ?: @"";
}

static NSString *VerdeMacJSONOrDescription(id value) {
    if (value == nil || value == [NSNull null]) return @"null";
    NSData *data = [NSJSONSerialization dataWithJSONObject:value options:NSJSONWritingFragmentsAllowed error:nil];
    if (data != nil) {
        NSString *json = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
        if (json != nil) return json;
    }
    return [value description] ?: @"null";
}

@implementation VerdeMacBrowser

- (instancetype)initWithWindow:(NSWindow *)window {
    self = [super init];
    if (self == nil) return nil;

    _window = window;
    _events = [NSMutableArray array];

    WKUserContentController *contentController = [[WKUserContentController alloc] init];
    WKUserScript *bridgeScript = [[WKUserScript alloc]
        initWithSource:@"(function(){"
                       "const bridge={postMessage:function(payload){window.webkit.messageHandlers.verde.postMessage(String(payload));}};"
                       "window.__VERDE_BROWSER_IPC__=bridge;"
                       "window.__VERDE_CEF_IPC__=bridge;"
                       "window.verde=bridge;"
                       "})();"
        injectionTime:WKUserScriptInjectionTimeAtDocumentStart
        forMainFrameOnly:NO];
    [contentController addUserScript:bridgeScript];
    [contentController addScriptMessageHandler:self name:@"verde"];

    WKWebViewConfiguration *configuration = [[WKWebViewConfiguration alloc] init];
    configuration.userContentController = contentController;
    configuration.preferences.javaScriptCanOpenWindowsAutomatically = YES;

    _container = [[NSView alloc] initWithFrame:NSZeroRect];
    _container.hidden = YES;
    _container.autoresizesSubviews = YES;

    _webView = [[WKWebView alloc] initWithFrame:_container.bounds configuration:configuration];
    _webView.navigationDelegate = self;
    _webView.autoresizingMask = NSViewWidthSizable | NSViewHeightSizable;
    [_container addSubview:_webView];

    NSView *contentView = window.contentView;
    [contentView addSubview:_container positioned:NSWindowAbove relativeTo:nil];
    [_webView addObserver:self forKeyPath:@"title" options:NSKeyValueObservingOptionNew context:NULL];
    [_webView addObserver:self forKeyPath:@"URL" options:NSKeyValueObservingOptionNew context:NULL];
    [_webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:@"about:blank"]]];
    return self;
}

- (void)dealloc {
    [self invalidate];
}

- (void)invalidate {
    if (self.invalidated) return;
    self.invalidated = YES;
    @try {
        [_webView removeObserver:self forKeyPath:@"title"];
        [_webView removeObserver:self forKeyPath:@"URL"];
    } @catch (__unused NSException *exception) {
    }
    _webView.navigationDelegate = nil;
    [_webView.configuration.userContentController removeScriptMessageHandlerForName:@"verde"];
    [_webView stopLoading];
    [_container removeFromSuperview];
}

- (void)queueEvent:(int)kind payload:(nullable NSString *)payload {
    if (self.invalidated) return;
    @synchronized(self.events) {
        [self.events addObject:[VerdeMacBrowserEvent eventWithKind:kind payload:payload]];
    }
}

- (void)setScreenX:(int)x y:(int)y width:(int)width height:(int)height {
    if (width <= 0 || height <= 0 || self.window == nil || self.window.contentView == nil) return;

    CGFloat scale = self.window.backingScaleFactor > 0.0 ? self.window.backingScaleFactor : 1.0;
    CGFloat pointHeight = MAX((CGFloat)height / scale, 1.0);
    CGFloat pointWidth = MAX((CGFloat)width / scale, 1.0);
    NSScreen *screen = self.window.screen ?: NSScreen.mainScreen;
    CGFloat screenTop = NSMaxY(screen.frame);
    NSRect targetScreenRect = NSMakeRect(
        (CGFloat)x / scale,
        screenTop - ((CGFloat)y / scale) - pointHeight,
        pointWidth,
        pointHeight);
    NSRect contentWindowRect = [self.window.contentView convertRect:self.window.contentView.bounds toView:nil];
    NSRect contentScreenRect = [self.window convertRectToScreen:contentWindowRect];

    self.container.frame = NSMakeRect(
        targetScreenRect.origin.x - contentScreenRect.origin.x,
        targetScreenRect.origin.y - contentScreenRect.origin.y,
        pointWidth,
        pointHeight);
    self.webView.frame = self.container.bounds;
}

- (void)userContentController:(WKUserContentController *)userContentController didReceiveScriptMessage:(WKScriptMessage *)message {
    (void)userContentController;
    NSString *payload = [message.body isKindOfClass:[NSString class]] ? (NSString *)message.body : VerdeMacJSONOrDescription(message.body);
    [self queueEvent:VERDE_MAC_BROWSER_EVENT_JS_MESSAGE payload:payload];
}

- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation {
    (void)webView;
    (void)navigation;
    [self queueEvent:VERDE_MAC_BROWSER_EVENT_DOCUMENT_LOADED payload:nil];
}

- (void)webView:(WKWebView *)webView didFailNavigation:(WKNavigation *)navigation withError:(NSError *)error {
    (void)webView;
    (void)navigation;
    [self queueEvent:VERDE_MAC_BROWSER_EVENT_FAILED payload:error.localizedDescription ?: @"Navigation failed"];
}

- (void)webView:(WKWebView *)webView didFailProvisionalNavigation:(WKNavigation *)navigation withError:(NSError *)error {
    (void)webView;
    (void)navigation;
    [self queueEvent:VERDE_MAC_BROWSER_EVENT_FAILED payload:error.localizedDescription ?: @"Navigation failed"];
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    (void)object;
    (void)change;
    (void)context;
    if ([keyPath isEqualToString:@"title"]) {
        if (self.webView.title.length > 0) {
            [self queueEvent:VERDE_MAC_BROWSER_EVENT_TITLE_CHANGED payload:self.webView.title];
        }
    } else if ([keyPath isEqualToString:@"URL"]) {
        NSString *absolute = self.webView.URL.absoluteString;
        if (absolute.length > 0) {
            [self queueEvent:VERDE_MAC_BROWSER_EVENT_NAVIGATED payload:absolute];
        }
    }
}

@end

void *verde_macos_webview_create(void *ns_window) {
    if (ns_window == NULL) return NULL;
    NSWindow *window = (__bridge NSWindow *)ns_window;
    VerdeMacBrowser *browser = [[VerdeMacBrowser alloc] initWithWindow:window];
    return (__bridge_retained void *)browser;
}

void verde_macos_webview_destroy(void *handle) {
    if (handle == NULL) return;
    VerdeMacBrowser *browser = (__bridge VerdeMacBrowser *)handle;
    [browser invalidate];
    CFBridgingRelease(handle);
}

int verde_macos_webview_show(void *handle) {
    if (handle == NULL) return 0;
    VerdeMacBrowser *browser = (__bridge VerdeMacBrowser *)handle;
    if (!browser.visible) {
        browser.visible = YES;
        [browser queueEvent:VERDE_MAC_BROWSER_EVENT_OPENED payload:nil];
    }
    browser.container.hidden = NO;
    [browser.window makeFirstResponder:browser.webView];
    return 1;
}

int verde_macos_webview_hide(void *handle) {
    if (handle == NULL) return 0;
    VerdeMacBrowser *browser = (__bridge VerdeMacBrowser *)handle;
    if (browser.visible) {
        browser.visible = NO;
        [browser queueEvent:VERDE_MAC_BROWSER_EVENT_CLOSED payload:nil];
    }
    browser.container.hidden = YES;
    return 1;
}

int verde_macos_webview_set_bounds(void *handle, int x, int y, int width, int height) {
    if (handle == NULL) return 0;
    VerdeMacBrowser *browser = (__bridge VerdeMacBrowser *)handle;
    [browser setScreenX:x y:y width:width height:height];
    return 1;
}

int verde_macos_webview_navigate(void *handle, const char *url) {
    if (handle == NULL || url == NULL) return 0;
    VerdeMacBrowser *browser = (__bridge VerdeMacBrowser *)handle;
    NSString *urlString = VerdeMacStringFromCString(url);
    NSURL *nsURL = [NSURL URLWithString:urlString];
    if (nsURL == nil) {
        [browser queueEvent:VERDE_MAC_BROWSER_EVENT_FAILED payload:@"Invalid URL"];
        return 0;
    }
    [browser.webView loadRequest:[NSURLRequest requestWithURL:nsURL]];
    return 1;
}

int verde_macos_webview_eval(void *handle, const char *js) {
    if (handle == NULL || js == NULL) return 0;
    VerdeMacBrowser *browser = (__bridge VerdeMacBrowser *)handle;
    NSString *script = VerdeMacStringFromCString(js);
    [browser.webView evaluateJavaScript:script completionHandler:^(id result, NSError *error) {
        if (error != nil) {
            [browser queueEvent:VERDE_MAC_BROWSER_EVENT_FAILED payload:error.localizedDescription ?: @"JavaScript evaluation failed"];
            return;
        }
        [browser queueEvent:VERDE_MAC_BROWSER_EVENT_EVAL_RESULT payload:VerdeMacJSONOrDescription(result)];
    }];
    return 1;
}

int verde_macos_webview_post_json(void *handle, const char *json) {
    if (handle == NULL || json == NULL) return 0;
    VerdeMacBrowser *browser = (__bridge VerdeMacBrowser *)handle;
    NSString *payload = VerdeMacStringFromCString(json);
    NSString *script = [NSString stringWithFormat:
        @"(function(){const payload=%@;window.dispatchEvent(new MessageEvent('verde-host-message',{data:payload}));})()",
        payload];
    [browser.webView evaluateJavaScript:script completionHandler:^(__unused id result, NSError *error) {
        if (error != nil) {
            [browser queueEvent:VERDE_MAC_BROWSER_EVENT_FAILED payload:error.localizedDescription ?: @"Browser JSON dispatch failed"];
        }
    }];
    return 1;
}

int verde_macos_webview_go_back(void *handle) {
    if (handle == NULL) return 0;
    VerdeMacBrowser *browser = (__bridge VerdeMacBrowser *)handle;
    if (browser.webView.canGoBack) [browser.webView goBack];
    return 1;
}

int verde_macos_webview_go_forward(void *handle) {
    if (handle == NULL) return 0;
    VerdeMacBrowser *browser = (__bridge VerdeMacBrowser *)handle;
    if (browser.webView.canGoForward) [browser.webView goForward];
    return 1;
}

int verde_macos_webview_reload(void *handle) {
    if (handle == NULL) return 0;
    VerdeMacBrowser *browser = (__bridge VerdeMacBrowser *)handle;
    [browser.webView reload];
    return 1;
}

int verde_macos_webview_focus(void *handle) {
    if (handle == NULL) return 0;
    VerdeMacBrowser *browser = (__bridge VerdeMacBrowser *)handle;
    [browser.window makeFirstResponder:browser.webView];
    return 1;
}

int verde_macos_webview_blur(void *handle) {
    if (handle == NULL) return 0;
    VerdeMacBrowser *browser = (__bridge VerdeMacBrowser *)handle;
    [browser.window makeFirstResponder:browser.window.contentView];
    return 1;
}

int verde_macos_webview_pop_event(void *handle, int *kind, char **payload) {
    if (handle == NULL || kind == NULL || payload == NULL) return 0;
    VerdeMacBrowser *browser = (__bridge VerdeMacBrowser *)handle;
    VerdeMacBrowserEvent *event = nil;
    @synchronized(browser.events) {
        if (browser.events.count == 0) return 0;
        event = browser.events.firstObject;
        [browser.events removeObjectAtIndex:0];
    }
    *kind = event.kind;
    *payload = event.payload != nil ? strdup(event.payload.UTF8String) : NULL;
    return 1;
}

void verde_macos_webview_free_string(char *value) {
    free(value);
}

File: packages/desktop/src/browser/platform/macos_wkwebview.swift
import AppKit
import Foundation
import WebKit

private enum VerdeMacBrowserEventKind: Int32 {
    case opened = 1
    case closed = 2
    case navigated = 3
    case titleChanged = 4
    case documentLoaded = 5
    case jsMessage = 6
    case evalResult = 7
    case failed = 8
}

private final class VerdeMacBrowserEvent {
    let kind: Int32
    let payload: String?

    init(kind: VerdeMacBrowserEventKind, payload: String? = nil) {
        self.kind = kind.rawValue
        self.payload = payload
    }
}

private final class VerdeFocusSinkView: NSView {
    override var acceptsFirstResponder: Bool { true }
}

private final class VerdeMacBrowser: NSObject, WKScriptMessageHandler, WKNavigationDelegate {
    private weak var window: NSWindow?
    private let focusSink: VerdeFocusSinkView
    private let container: NSView
    private let webView: WKWebView
    private var events: [VerdeMacBrowserEvent] = []
    private let eventLock = NSLock()
    private var titleObservation: NSKeyValueObservation?
    private var urlObservation: NSKeyValueObservation?
    private var visible = false
    private var invalidated = false

    init(window: NSWindow) {
        self.window = window
        Self.configureForegroundApp(window: window)

        let contentController = WKUserContentController()
        let bridgeScript = """
        (function(){
          const bridge={postMessage:function(payload){window.webkit.messageHandlers.verde.postMessage(String(payload));}};
          window.__VERDE_BROWSER_IPC__=bridge;
          window.__VERDE_CEF_IPC__=bridge;
          window.verde=bridge;
        })();
        """
        contentController.addUserScript(WKUserScript(source: bridgeScript, injectionTime: .atDocumentStart, forMainFrameOnly: false))

        let configuration = WKWebViewConfiguration()
        configuration.userContentController = contentController
        configuration.preferences.javaScriptCanOpenWindowsAutomatically = true

        self.focusSink = VerdeFocusSinkView(frame: .zero)
        self.container = NSView(frame: .zero)
        self.container.isHidden = true
        self.container.autoresizesSubviews = true
        self.webView = WKWebView(frame: .zero, configuration: configuration)
        self.webView.autoresizingMask = [.width, .height]

        super.init()

        contentController.add(self, name: "verde")
        self.webView.navigationDelegate = self
        self.container.addSubview(self.webView)
        window.contentView?.addSubview(self.focusSink, positioned: .below, relativeTo: nil)
        window.contentView?.addSubview(self.container, positioned: .above, relativeTo: nil)

        self.titleObservation = self.webView.observe(\.title, options: [.new]) { [weak self] webView, _ in
            guard let self, !self.invalidated, let title = webView.title, !title.isEmpty else { return }
            self.queueEvent(.titleChanged, payload: title)
        }
        self.urlObservation = self.webView.observe(\.url, options: [.new]) { [weak self] webView, _ in
            guard let self, !self.invalidated, let absolute = webView.url?.absoluteString, !absolute.isEmpty else { return }
            self.queueEvent(.navigated, payload: absolute)
        }
        self.webView.load(URLRequest(url: URL(string: "about:blank")!))
    }

    deinit {
        invalidate()
    }

    func invalidate() {
        if invalidated { return }
        invalidated = true
        titleObservation?.invalidate()
        titleObservation = nil
        urlObservation?.invalidate()
        urlObservation = nil
        webView.navigationDelegate = nil
        webView.configuration.userContentController.removeScriptMessageHandler(forName: "verde")
        webView.stopLoading()
        focusSink.removeFromSuperview()
        container.removeFromSuperview()
    }

    func queueEvent(_ kind: VerdeMacBrowserEventKind, payload: String? = nil) {
        if invalidated { return }
        eventLock.lock()
        events.append(VerdeMacBrowserEvent(kind: kind, payload: payload))
        eventLock.unlock()
    }

    func popEvent() -> VerdeMacBrowserEvent? {
        eventLock.lock()
        defer { eventLock.unlock() }
        if events.isEmpty { return nil }
        return events.removeFirst()
    }

    func setBounds(x: Int32, y: Int32, width: Int32, height: Int32) {
        guard width > 0, height > 0, let window, let contentView = window.contentView else { return }
        let scale = window.backingScaleFactor > 0 ? window.backingScaleFactor : 1.0
        let pointWidth = max(CGFloat(width) / scale, 1.0)
        let pointHeight = max(CGFloat(height) / scale, 1.0)
        let screen = window.screen ?? NSScreen.main
        let screenTop = screen?.frame.maxY ?? 0
        let targetScreenRect = NSRect(x: CGFloat(x) / scale, y: screenTop - (CGFloat(y) / scale) - pointHeight, width: pointWidth, height: pointHeight)
        let contentWindowRect = contentView.convert(contentView.bounds, to: nil)
        let contentScreenRect = window.convertToScreen(contentWindowRect)

        container.frame = NSRect(
            x: targetScreenRect.origin.x - contentScreenRect.origin.x,
            y: targetScreenRect.origin.y - contentScreenRect.origin.y,
            width: pointWidth,
            height: pointHeight
        )
        webView.frame = container.bounds
    }

    func show() {
        Self.configureForegroundApp(window: window)
        if !visible {
            visible = true
            queueEvent(.opened)
        }
        container.isHidden = false
    }

    func hide() {
        if visible {
            visible = false
            queueEvent(.closed)
        }
        container.isHidden = true
    }

    func navigate(_ urlString: String) -> Bool {
        guard let url = URL(string: urlString) else {
            queueEvent(.failed, payload: "Invalid URL")
            return false
        }
        webView.load(URLRequest(url: url))
        return true
    }

    func eval(_ script: String) {
        webView.evaluateJavaScript(script) { [weak self] result, error in
            guard let self, !self.invalidated else { return }
            if let error {
                self.queueEvent(.failed, payload: error.localizedDescription)
                return
            }
            self.queueEvent(.evalResult, payload: Self.jsonOrDescription(result))
        }
    }

    func postJson(_ json: String) {
        let script = "(function(){const payload=\(json);window.dispatchEvent(new MessageEvent('verde-host-message',{data:payload}));})()"
        webView.evaluateJavaScript(script) { [weak self] _, error in
            guard let self, !self.invalidated, let error else { return }
            self.queueEvent(.failed, payload: error.localizedDescription)
        }
    }

    func goBack() {
        if webView.canGoBack { webView.goBack() }
    }

    func goForward() {
        if webView.canGoForward { webView.goForward() }
    }

    func reload() {
        webView.reload()
    }

    func focus() {
        guard let window else { return }
        Self.configureForegroundApp(window: window)
        window.makeFirstResponder(webView)
    }

    func blur() {
        window?.makeFirstResponder(focusSink)
    }

    func hasFocus() -> Bool {
        guard let window, window.isKeyWindow, NSApp.isActive, let responder = window.firstResponder else { return false }
        if responder === webView { return true }
        if let view = responder as? NSView {
            return view === webView || view.isDescendant(of: webView)
        }
        return false
    }

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        let payload = (message.body as? String) ?? Self.jsonOrDescription(message.body)
        queueEvent(.jsMessage, payload: payload)
    }

    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        queueEvent(.documentLoaded)
    }

    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        queueEvent(.failed, payload: error.localizedDescription)
    }

    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        queueEvent(.failed, payload: error.localizedDescription)
    }

    private static func jsonOrDescription(_ value: Any?) -> String {
        guard let value else { return "null" }
        if JSONSerialization.isValidJSONObject(value),
           let data = try? JSONSerialization.data(withJSONObject: value, options: [.fragmentsAllowed]),
           let json = String(data: data, encoding: .utf8) {
            return json
        }
        if let string = value as? String {
            return string
        }
        return String(describing: value)
    }

    static func configureForegroundApp(window: NSWindow? = nil) {
        NSApp.setActivationPolicy(.regular)
        NSApp.unhide(nil)

        if let window {
            window.canHide = false
            if window.isMiniaturized {
                window.deminiaturize(nil)
            }
            window.setIsVisible(true)
            window.orderFrontRegardless()
            window.makeKeyAndOrderFront(nil)
            window.makeMain()
        }

        NSRunningApplication.current.activate(options: [.activateAllWindows, .activateIgnoringOtherApps])
        NSApp.activate(ignoringOtherApps: true)
    }

    func appKitDiagnostics() -> String {
        return Self.appKitDiagnostics(window: window, webView: webView, container: container)
    }

    static func appKitDiagnostics(window: NSWindow?, webView: WKWebView? = nil, container: NSView? = nil) -> String {
        let responderDescription: String
        if let responder = window?.firstResponder {
            responderDescription = String(describing: type(of: responder))
        } else {
            responderDescription = "null"
        }

        var payload: [String: Any] = [
            "appActive": NSApp.isActive,
            "activationPolicy": NSApp.activationPolicy().rawValue,
            "appWindowCount": NSApp.windows.count,
            "runningApplicationActive": NSRunningApplication.current.isActive,
            "hasKeyWindow": NSApp.keyWindow != nil,
            "hasMainWindow": NSApp.mainWindow != nil,
        ]

        if let window {
            payload["windowPresent"] = true
            payload["windowTitle"] = window.title
            payload["windowIsVisible"] = window.isVisible
            payload["windowIsKey"] = window.isKeyWindow
            payload["windowIsMain"] = window.isMainWindow
            payload["windowCanBecomeKey"] = window.canBecomeKey
            payload["windowCanBecomeMain"] = window.canBecomeMain
            payload["windowIsMiniaturized"] = window.isMiniaturized
            payload["windowCanHide"] = window.canHide
            payload["windowStyleMask"] = window.styleMask.rawValue
            payload["windowLevel"] = window.level.rawValue
            payload["windowFrame"] = [
                "x": window.frame.origin.x,
                "y": window.frame.origin.y,
                "width": window.frame.size.width,
                "height": window.frame.size.height,
            ]
            payload["contentViewSubviewCount"] = window.contentView?.subviews.count ?? 0
            payload["firstResponder"] = responderDescription
        } else {
            payload["windowPresent"] = false
        }

        if let webView {
            payload["webViewWindowAttached"] = webView.window != nil
            payload["webViewHidden"] = webView.isHidden
            payload["webViewFrame"] = [
                "x": webView.frame.origin.x,
                "y": webView.frame.origin.y,
                "width": webView.frame.size.width,
                "height": webView.frame.size.height,
            ]
        }

        if let container {
            payload["containerWindowAttached"] = container.window != nil
            payload["containerHidden"] = container.isHidden
            payload["containerFrame"] = [
                "x": container.frame.origin.x,
                "y": container.frame.origin.y,
                "width": container.frame.size.width,
                "height": container.frame.size.height,
            ]
        }

        guard JSONSerialization.isValidJSONObject(payload),
              let data = try? JSONSerialization.data(withJSONObject: payload, options: [.sortedKeys]),
              let json = String(data: data, encoding: .utf8) else {
            return "{}"
        }
        return json
    }
}

private func onMain<T>(_ body: @escaping () -> T) -> T {
    if Thread.isMainThread {
        return body()
    }
    return DispatchQueue.main.sync(execute: body)
}

private func stringFromCString(_ value: UnsafePointer<CChar>?) -> String {
    guard let value else { return "" }
    return String(cString: value)
}

private func browserFromOpaque(_ handle: UnsafeMutableRawPointer?) -> VerdeMacBrowser? {
    guard let handle else { return nil }
    return Unmanaged<VerdeMacBrowser>.fromOpaque(handle).takeUnretainedValue()
}

@_cdecl("verde_macos_webview_create")
public func verde_macos_webview_create(_ nsWindow: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let nsWindow else { return nil }
    return onMain {
        let window = Unmanaged<NSWindow>.fromOpaque(nsWindow).takeUnretainedValue()
        let browser = VerdeMacBrowser(window: window)
        return Unmanaged.passRetained(browser).toOpaque()
    }
}

@_cdecl("verde_macos_app_configure_foreground")
public func verde_macos_app_configure_foreground() {
    onMain {
        VerdeMacBrowser.configureForegroundApp()
    }
}

@_cdecl("verde_macos_webview_appkit_diagnostics")
public func verde_macos_webview_appkit_diagnostics(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    onMain {
        let json: String
        if let browser = browserFromOpaque(handle) {
            json = browser.appKitDiagnostics()
        } else {
            json = VerdeMacBrowser.appKitDiagnostics(window: nil)
        }
        return strdup(json)
    }
}

@_cdecl("verde_macos_webview_destroy")
public func verde_macos_webview_destroy(_ handle: UnsafeMutableRawPointer?) {
    guard let handle else { return }
    onMain {
        let browser = Unmanaged<VerdeMacBrowser>.fromOpaque(handle).takeRetainedValue()
        browser.invalidate()
    }
}

@_cdecl("verde_macos_webview_show")
public func verde_macos_webview_show(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    onMain {
        guard let browser = browserFromOpaque(handle) else { return 0 }
        browser.show()
        return 1
    }
}

@_cdecl("verde_macos_webview_hide")
public func verde_macos_webview_hide(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    onMain {
        guard let browser = browserFromOpaque(handle) else { return 0 }
        browser.hide()
        return 1
    }
}

@_cdecl("verde_macos_webview_set_bounds")
public func verde_macos_webview_set_bounds(_ handle: UnsafeMutableRawPointer?, _ x: Int32, _ y: Int32, _ width: Int32, _ height: Int32) -> Int32 {
    onMain {
        guard let browser = browserFromOpaque(handle) else { return 0 }
        browser.setBounds(x: x, y: y, width: width, height: height)
        return 1
    }
}

@_cdecl("verde_macos_webview_navigate")
public func verde_macos_webview_navigate(_ handle: UnsafeMutableRawPointer?, _ url: UnsafePointer<CChar>?) -> Int32 {
    onMain {
        guard let browser = browserFromOpaque(handle), let url else { return 0 }
        return browser.navigate(stringFromCString(url)) ? 1 : 0
    }
}

@_cdecl("verde_macos_webview_eval")
public func verde_macos_webview_eval(_ handle: UnsafeMutableRawPointer?, _ js: UnsafePointer<CChar>?) -> Int32 {
    onMain {
        guard let browser = browserFromOpaque(handle), let js else { return 0 }
        browser.eval(stringFromCString(js))
        return 1
    }
}

@_cdecl("verde_macos_webview_post_json")
public func verde_macos_webview_post_json(_ handle: UnsafeMutableRawPointer?, _ json: UnsafePointer<CChar>?) -> Int32 {
    onMain {
        guard let browser = browserFromOpaque(handle), let json else { return 0 }
        browser.postJson(stringFromCString(json))
        return 1
    }
}

@_cdecl("verde_macos_webview_go_back")
public func verde_macos_webview_go_back(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    onMain {
        guard let browser = browserFromOpaque(handle) else { return 0 }
        browser.goBack()
        return 1
    }
}

@_cdecl("verde_macos_webview_go_forward")
public func verde_macos_webview_go_forward(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    onMain {
        guard let browser = browserFromOpaque(handle) else { return 0 }
        browser.goForward()
        return 1
    }
}

@_cdecl("verde_macos_webview_reload")
public func verde_macos_webview_reload(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    onMain {
        guard let browser = browserFromOpaque(handle) else { return 0 }
        browser.reload()
        return 1
    }
}

@_cdecl("verde_macos_webview_focus")
public func verde_macos_webview_focus(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    onMain {
        guard let browser = browserFromOpaque(handle) else { return 0 }
        browser.focus()
        return 1
    }
}

@_cdecl("verde_macos_webview_blur")
public func verde_macos_webview_blur(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    onMain {
        guard let browser = browserFromOpaque(handle) else { return 0 }
        browser.blur()
        return 1
    }
}

@_cdecl("verde_macos_webview_has_focus")
public func verde_macos_webview_has_focus(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    onMain {
        guard let browser = browserFromOpaque(handle) else { return 0 }
        return browser.hasFocus() ? 1 : 0
    }
}

@_cdecl("verde_macos_webview_pop_event")
public func verde_macos_webview_pop_event(_ handle: UnsafeMutableRawPointer?, _ kind: UnsafeMutablePointer<Int32>?, _ payload: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?) -> Int32 {
    guard let kind, let payload else { return 0 }
    return onMain {
        guard let browser = browserFromOpaque(handle), let event = browser.popEvent() else { return 0 }
        kind.pointee = event.kind
        payload.pointee = event.payload.flatMap { strdup($0) }
        return 1
    }
}

@_cdecl("verde_macos_webview_free_string")
public func verde_macos_webview_free_string(_ value: UnsafeMutablePointer<CChar>?) {
    free(value)
}
